pkg/virtctl/vmexport/vmexport.go

```diff
@@ -383,67 +483,124 @@ func downloadVolume(client kubecli.KubevirtClient, vmexport *exportv1.VirtualMac
 		return err
 	}
 
-	fmt.Println("Download finished succesfully")
+	// Prevent this output ending up in the stdout
+	if vmeInfo.OutputFile != "" {
+		fmt.Println("Download finished succesfully")
+	}
 	return nil
 }
 
+func replaceUrlWithServiceUrl(manifestUrl string, vmeInfo *VMExportInfo) (string, error) {
+	// Replace internal URL with specified URL
+	manUrl, err := url.Parse(manifestUrl)
+	if err != nil {
+		return "", err
+	}
+	if vmeInfo.ServiceURL != "" {
+		manUrl.Host = vmeInfo.ServiceURL
+	}
+	return manUrl.String(), nil
+}
+
 // GetUrlFromVirtualMachineExport inspects the VirtualMachineExport status to fetch the extected URL
 func GetUrlFromVirtualMachineExport(vmexport *exportv1.VirtualMachineExport, vmeInfo *VMExportInfo) (string, error) {
-	var downloadUrl string
+	var (
+		downloadUrl string
+		err         error
+		links       *exportv1.VirtualMachineExportLink
+	)
 
-	if vmexport.Status.Links == nil || vmexport.Status.Links.External == nil || len(vmexport.Status.Links.External.Volumes) <= 0 {
-		return "", fmt.Errorf("Unable to access the volume info from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
+	if vmeInfo.ServiceURL == "" {
+		if vmexport.Status.Links == nil || vmexport.Status.Links.External == nil {
+			return "", fmt.Errorf("unable to access the manifest info from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
+		}
+		links = vmexport.Status.Links.External
+	} else {
+		if vmexport.Status.Links == nil || vmexport.Status.Links.Internal == nil {
+			return "", fmt.Errorf("unable to access the volume info from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
+		}
+		links = vmexport.Status.Links.Internal
 	}
-
-	volumeNumber := len(vmexport.Status.Links.External.Volumes)
+	if len(links.Volumes) <= 0 {
+		return "", fmt.Errorf("unable to access the volume info from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
+	}
+	volumeNumber := len(links.Volumes)
 	if volumeNumber > 1 && vmeInfo.VolumeName == "" {
-		return "", fmt.Errorf("Detected more than one downloadable volume in '%s/%s' VirtualMachineExport: Select the expected volume using the --volume flag", vmexport.Namespace, vmexport.Name)
+		return "", fmt.Errorf("detected more than one downloadable volume in '%s/%s' VirtualMachineExport: Select the expected volume using the --volume flag", vmexport.Namespace, vmexport.Name)
 	}
-
-	for _, exportVolume := range vmexport.Status.Links.External.Volumes {
+	for _, exportVolume := range links.Volumes {
 		// Access the requested volume
 		if volumeNumber == 1 || exportVolume.Name == vmeInfo.VolumeName {
 			for _, format := range exportVolume.Formats {
 				// We always attempt to find and get the compressed file URL, so we only break the loop when one is found
 				if format.Format == exportv1.KubeVirtGz || format.Format == exportv1.ArchiveGz {
-					downloadUrl = format.Url
+					downloadUrl, err = replaceUrlWithServiceUrl(format.Url, vmeInfo)
```

Comment:
I feel that `replaceUrlWithServiceUrl` sounds too specific for something that we are only doing when `ServiceURL` is set. I would use something more generic like `resolveExportUrl`, wdyt?

Reply:
naming is a hard thing. I personally like `replaceUrlWithServiceUrl` better than your suggestion. It is very explicit about the limited scope of what it is doing. Maybe `replaceDomainWithServiceUrl`. Since essentially we are replacing the domain part of the volume URL with the service-url